packages/unified-selection/src/unified-selection/HiliteSetProvider.ts

```diff
@@ -0,0 +1,338 @@
+import { ECSqlBinding, IECSqlQueryExecutor } from "./queries/ECSqlCore";
+import { SelectableInstanceKey, Selectables } from "./Selectable";
+
+/**
+ * A set of model, subcategory and element ids that can be used for specifying hilite.
+ * @public
+ */
+export interface HiliteSet {
+  models: string[];
+  subCategories: string[];
+  elements: string[];
+}
+
+/**
+ * Props for creating a `HiliteSetProvider` instance.
+ * @public
+ */
+export interface HiliteSetProviderProps {
+  queryExecutor: IECSqlQueryExecutor;
+  metadataProvider: IMetadataProvider;
+}
+
+/**
+ * ECSQL based provider which determines what `HiliteSet`
+ * should be hilited based on the supplied `Selectables`.
+ * @public
+ */
+export class HiliteSetProvider {
+  private _queryExecutor: IECSqlQueryExecutor;
+  private _metadataProvider: IMetadataProvider;
+  // Map between a class name and its type
+  private _classRelationCache: Map<string, InstanceIdType>;
+  private _hasFunctionalSchema?: boolean;
+
+  private constructor(props: HiliteSetProviderProps) {
+    this._queryExecutor = props.queryExecutor;
+    this._metadataProvider = props.metadataProvider;
+    this._classRelationCache = new Map<string, InstanceIdType>();
+  }
+
+  /**
+   * Create a hilite set provider.
+   */
+  public static create(props: HiliteSetProviderProps) {
+    return new HiliteSetProvider(props);
+  }
+
+  /**
+   * Get hilite set for supplied `Selectables`.
+   */
+  public async getHiliteSet(selection: Selectables): Promise<HiliteSet> {
+    // Map between element ID types and EC instance IDs
+    const keysByType = new Map<InstanceIdType, string[]>();
+    if (!this._hasFunctionalSchema) {
+      try {
+        this._hasFunctionalSchema = !!(await this._metadataProvider.getSchema("Functional"));
+      } catch {
+        this._hasFunctionalSchema = false;
+      }
+    }
+
+    for (const entry of selection.instanceKeys) {
+      for (const key of entry[1]) {
+        await this.addKeyByType({ className: entry[0], id: key }, keysByType);
+      }
+    }
+
+    for (const entry of selection.custom) {
+      for await (const key of entry[1].loadInstanceKeys()) {
+        await this.addKeyByType(key, keysByType);
+      }
+    }
+
+    return {
+      models: await this.getHilitedModels(keysByType),
+      subCategories: await this.getHilitedSubCategories(keysByType),
+      elements: await this.getHilitedElements(keysByType),
+    };
+  }
+
+  private async addKeyByType(key: SelectableInstanceKey, map: Map<string, string[]>) {
+    const keyType = await this.getType(key);
+    let array = map.get(keyType);
+    if (!array) {
+      array = [];
+    }
+
+    array.push(key.id);
+    map.set(keyType, array);
+  }
+
+  private async getType(key: SelectableInstanceKey): Promise<InstanceIdType> {
+    const cachedType = this._classRelationCache.get(key.className);
+    if (cachedType) {
+      return cachedType;
+    }
+
+    const keyClass = await this.getClass(key.className);
```

Comment:
I think this could be moved to `checkType`

Reply:
That would cause the `getClass` function to be called every time the type is checked. Not sure how expensive the call is but I think it is better as it is currently.